### PR TITLE
Add file upload exercises to exam mode

### PR DESCRIPTION
**Pull request by SimonEntholzer**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

The exam mode currently doesn't have the file-upload-exercise type.

### Description
In this PR we add the file-upload exercise to the exam mode. Instructors can now create file upload exercises in the exercise groups page.
image
image

Assessment works already too, but only for one correction round. Assessment in the second correction round will be added in the next PR.

### Steps for Testing

1. Create a new exam, add a new exercise group and add a file-upload-exercise.
2. Test the exam exercise first in a test run, and then with a real exam.
image

3. upload a file, and click on upload. Now the uploaded file should be visible below, and you should be able to download it again.

4. upload a different file, and test if it still works.
5. hand in the exam
6. In the summary at the end you should be able to download the file again.
 

### Test Coverage

file-upload-exam-submission.component.ts: 94.92%

### CODE REVIEW

File: src/main/webapp/app/exam/participate/exercises/file-upload/file-upload-exam-submission.component.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+import { Result } from 'app/entities/result.model';
+import { ExamSubmissionComponent } from 'app/exam/participate/exercises/exam-submission.component';
+import { Exercise, IncludedInOverallScore } from 'app/entities/exercise.model';
+import { Submission } from 'app/entities/submission.model';
+
+@Component({
+    selector: 'jhi-file-upload-submission-exam',
+    templateUrl: './file-upload-exam-submission.component.html',
+    providers: [{ provide: ExamSubmissionComponent, useExisting: FileUploadExamSubmissionComponent }],
+    styleUrls: ['./file-upload-exam-submission.component.scss'],
```
Comment (fde312) [inline src/main/webapp/app/exam/participate/exercises/file-upload/file-upload-exam-submission.component.ts:25]:
```suggestion

```

**Review by JohannesWeiss — APPROVED** [review]
Tested on ts1 and everything worked fine so far, even the (one round) assessment.

**Review by airelawaleria — DISMISSED** [review]
All worked fine with test runs and exam mode. 

The only one small thing is the button arrangement, maybe it would be nice to place the quiz button under the file upload

image

**Review by EneaGore — APPROVED** [review]
Looks fine on ts1.

**Comment by SimonEntholzer** [comment]
> All worked fine with test runs and exam mode.
> 
> The only one small thing is the button arrangement, maybe it would be nice to place the quiz button under the file upload
> 
> image

It's intended that the button is next to the quiz button, and not on top of each other. Later we will maybe add an import button. I'm also not sure why it is displayed at the top. (I guess you use chrome?) Normally it should get displayed in the same row as all the other 'add' buttons.

**Review by marvinchimeka — DISMISSED** [review]
Tested on ts1 with @simonzilker  . Works fine. Could not see any issues.

**Review by simonzilker — APPROVED** [review]
Looks good :)

**Review by philgabr — APPROVED** [review]
Approve :+1:

**Review by valeriebucher — DISMISSED** [review]

**Review by DenizShabani — DISMISSED** [review]

File: src/main/resources/config/application-dev.yml [hunk 2]
```diff
@@ -60,7 +60,7 @@ spring:
         localInstances: true
 server:
     port: 8080
-    url: https://artemislocal.ase.in.tum.de
+    url: http://docker.for.mac.host.internal:8080
```
Comment (FrankeLukas) [inline src/main/resources/config/application-dev.yml:63]: This likely needs to be reverted.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseParticipationService.java [hunk 3]
```diff
@@ -339,12 +339,17 @@ public void unlockStudentRepository(ProgrammingExercise programmingExercise, Pro
...
+                    gitService.stashChanges(repo);
+                }
+                else {
+                    log.warn("Cannot stash student repository for participation " + participation.getId() + " because the due date has not passed yet!");
+                }
+
             }
             catch (InterruptedException | GitAPIException e) {
-                log.error("Stashing student repository for participation " + participation.getId() + " in exercise '" + programmingExercise.getTitle()
+                log.error("Stashing student repository for participation" + participation.getId() + " in exercise '" + programmingExercise.getTitle()
```
Comment (FrankeLukas) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseParticipationService.java:352]: Is that change intentional? That will cramp `participation` and the id together.

Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseParticipationService.java:352]: No somehow this is reverts a change that was done previously

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseParticipationService.java [hunk 4]
```diff
@@ -344,7 +344,7 @@ public void stashChangesInStudentRepositoryAfterDueDateHasPassed(ProgrammingExer
                 gitService.stashChanges(repo);
             }
             catch (InterruptedException | GitAPIException e) {
-                log.error("Stashing student repository for participation " + participation.getId() + " in exercise '" + programmingExercise.getTitle()
+                log.error("Stashing student repository for participation" + participation.getId() + " in exercise '" + programmingExercise.getTitle()
```
Comment (fde312) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseParticipationService.java:347]:
```suggestion
                log.error("Stashing student repository for participation " + participation.getId() + " in exercise '" + programmingExercise.getTitle()
```

**Review by derLalla — APPROVED** [review]
code looks good!

File: src/main/webapp/app/exercises/shared/exam-exercise-row-buttons/exam-exercise-row-buttons.component.html [hunk 5]
```diff
@@ -96,6 +96,7 @@
 
     <a
         *ngIf="course.isAtLeastInstructor && exercise.type !== exerciseType.QUIZ"
+        type="submit"
```
Comment (SimonEntholzer) [inline src/main/webapp/app/exercises/shared/exam-exercise-row-buttons/exam-exercise-row-buttons.component.html:99]:
```suggestion
```
not needed for links anymore

File: src/main/webapp/app/exam/participate/exam-participation.component.ts [hunk 6]
```diff
@@ -270,7 +273,7 @@ export class ExamParticipationComponent implements OnInit, OnDestroy, ComponentC
         // auto save of submission if there are changes
         this.autoSaveInterval = window.setInterval(() => {
             this.autoSaveTimer++;
-            if (this.autoSaveTimer >= 30 && !this.isOver()) {
+            if (this.autoSaveTimer >= 300 && !this.isOver()) {
```
Comment (SimonEntholzer) [inline src/main/webapp/app/exam/participate/exam-participation.component.ts:276]:
```suggestion
            if (this.autoSaveTimer >= 30 && !this.isOver()) {
```

**Review by FrankeLukas — APPROVED** [review]
Tested in testing session.

**Review by derLalla — APPROVED** [review]
Tested in testing session

**Review by kloessst — APPROVED** [review]
Worked fine in the test run. Haven't done a code review though
